sam_view.c

```diff
@@ -679,24 +702,92 @@ typedef struct bam2fq_opts {
     char *fnse;
     char *fnr[3];
     char *fn_input; // pointer to input filename in argv do not free
-    bool has12, use_oq, copy_tags;
+    bool has12, has12always, use_oq, copy_tags;
     int flag_on, flag_off, flag_alloff;
     sam_global_args ga;
     fastfile filetype;
     int def_qual;
+    char *barcode_tag;
+    char *quality_tag;
+    char *index_file[2];
+    char *index_format;
 } bam2fq_opts_t;
 
 typedef struct bam2fq_state {
     samFile *fp;
     FILE *fpse;
     FILE *fpr[3];
+    FILE *fpi[2];
     bam_hdr_t *h;
     bool has12, use_oq, copy_tags;
     int flag_on, flag_off, flag_alloff;
     fastfile filetype;
     int def_qual;
 } bam2fq_state_t;
 
+/*
+ * get and decode the read from a BAM record
+ * Why on earth isn't this (and the next few functions) part of htslib?
+ * The whole point of an API is that we should not have to know the internal structure...
+ */
+
+/*
+ * reverse a string in place
+ (stolen from http://stackoverflow.com/questions/8534274/is-the-strrev-function-not-available-in-linux)
+ */
+static char *reverse(char *str)
+{
+    int i = strlen(str)-1,j=0;
+    char ch;
+    while (i>j) {
+        ch = str[i];
+        str[i]= str[j];
+        str[j] = ch;
+        i--;
+        j++;
+    }
+    return str;
+}
+
+/* return the read, reverse complemented if necessary */
+static char *get_read(const bam1_t *rec)
+{
+    int len = rec->core.l_qseq + 1;
+    char *read = calloc(1, kroundup32(len));
```

Comment:
I was following Matt's suggestion at https://matt.sh/howto-c
It does, as you say, give you a nul termination. Also, I *like* calloc()

Looking at the code carefully, I agree that kroundup32() is not strictly necessary here, but I have in the past found problems with a malloc'd string when using strcpy() when it was not rounded up and optimisation was on. It took me some time to track that one down, so I'm understandable cautious now.

Checking the return value: well spotted.